types/heatmap.js/index.d.ts

```diff
@@ -1,127 +1,291 @@
-// Type definitions for heatmap.js v2.0
+// Type definitions for heatmap.js 2.0
 // Project: https://github.com/pa7/heatmap.js/
 // Definitions by: Yang Guan <https://github.com/lookuptable>
+//                 Rhys van der Waerden <https://github.com/rhys-vdw>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
-// TypeScript Version: 2.3
+// TypeScript Version: 2.4
 
-import * as Leaflet from "leaflet";
+export as namespace h337;
 
-declare global {
+/*
+ * Create a heatmap instance. A Heatmap can be customized with the configObject.
+ */
+export function create<
```

Comment:
Could this be:
```ts
export function create<
    K extends string = "x" | "y" | "value",
    O extends { [key in K]: number } = Record<K, number>
>(
    configObject: HeatmapConfiguration<K, O>
): Heatmap<K, O>
```?
It would mean that you don't need a `DataPoint` type since `O` serves that purpose (and defaults to `Record<K, number>` which is basically what `DataPoint` was).
CC @weswigham @sandersn to see if this is the best solution

Reply:
I think your CCs here wont have triggered due to them accidentally being included in a code block.

Reply:
Is the `O` parameter necessary? Only the `K` type is referenced by the config... So it would be:

```ts
interface HeatmapConfiguration<K extends string = 'x' | 'y' | 'value'> {
  // ...
  xField?: K,
  yField?: K,
  valueField?: K,
}
```

But the `O` could be used to constrain the input to `Heatmap#addData`, `Heatmap#setData` which would prevent accidental insertions of other types that also implement the same interface as `Record<K, number>`...

eg.

```ts
interface MyDataPoint {
  foo: string, x: number, y: number, value: number
}
const heatmap = h337.create<'x' | 'y' | 'value', MyDataPoint>(config)
heatmap.addData({ x: 0, y: 1, value: 5 }) // ERROR! Expected an instance of `MyDataPoint`
```

Note that `getData` doesn't [return the same object you insert](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/21597/files#diff-091758ba3ab532fd30e93538bdcf533dR84), so there's no real value beyond the above.

Is that the idea? Or is more about just having fewer concepts for simplicity?

Reply:
I think I get it -- `O` ensures they keys are mapped to `number`, right?

Reply:
If `getData` doesn't return the object you put in it may be better to omit `O` and just use `Record<K, number>` directly.